src/core/net/srp_client.cpp

```diff
@@ -1027,12 +1067,30 @@ Error Client::AppendHostDescriptionInstruction(Message &aMessage, Info &aInfo) c
     rr.SetTtl(mLeaseInterval);
     rr.SetLength(sizeof(Ip6::Address));
 
-    for (uint8_t index = 0; index < mHostInfo.GetNumAddresses(); index++)
+    if (mHostInfo.IsAutoAddressEnabled())
     {
-        SuccessOrExit(error = AppendHostName(aMessage, aInfo));
-        SuccessOrExit(error = aMessage.Append(rr));
-        SuccessOrExit(error = aMessage.Append(mHostInfo.GetAddress(index)));
-        aInfo.mRecordCount++;
+        for (const Ip6::Netif::UnicastAddress &unicastAddress : Get<ThreadNetif>().GetUnicastAddresses())
+        {
+            if (unicastAddress.GetAddress().IsLinkLocal() || unicastAddress.GetAddress().GetIid().IsLocator())
+            {
+                continue;
+            }
+
+            SuccessOrExit(error = AppendHostName(aMessage, aInfo));
+            SuccessOrExit(error = aMessage.Append(rr));
+            SuccessOrExit(error = aMessage.Append(unicastAddress.GetAddress()));
+            aInfo.mRecordCount++;
+        }
+    }
+    else
+    {
+        for (uint8_t index = 0; index < mHostInfo.GetNumAddresses(); index++)
```

Comment:
Would it be somehow beneficial to implement the `HostInfo` list of addresses as a `LinkedList` to avoid this code duplication?

Reply:
I am not sure it can help much (in terms of allowing us to share same loop/code here) since the host addresses provided by user won't be same type as `Ip6::Netif::UnicastAddress` struct.

Also it would require a change in public SRP client API `otSrpClientSetHostAddresses()` to accept a linked-list instead of an array of addresses. In general, I think it is a good idea to try to minimize changes to previous APIs (to keep it backward compatible) unless we have a good justification.